src/QuickCheckVEngine/Main.hs

```diff
@@ -176,40 +177,45 @@ main :: IO ()
   rawArgs <- getArgs
-  (flags, leftover) <- commandOpts rawArgs
+  (flags, _) <- commandOpts rawArgs
   when (optVerbosity flags > 1) $ print flags
   let archDesc = arch flags
   -- initialize model and implementation sockets
   addrA <- resolve (impAIP flags) (impAPort flags)
   addrB <- resolve (impBIP flags) (impBPort flags)
   socA <- open "implementation-A" addrA
   socB <- open "implementation-B" addrB
-  sendDIIPacket socA diiEnd
-  _ <- recvRVFITrace socA False
-  sendDIIPacket socB diiEnd
-  _ <- recvRVFITrace socB False
+  socATraceVer <- rvfiNegotiateVersion socA "implementation A" (optVerbosity flags)
+  socBTraceVer <- rvfiNegotiateVersion socB "implementation B" (optVerbosity flags)
+  -- XXX: Data constructor not in scope:
+  -- let connectionA = RvfiDiiConnection socA socATraceVer "implementation A" (optVerbosity flags)
+  -- let connectionB = RvfiDiiConnection socB socBTraceVer "implementation B" (optVerbosity flags)
+  let connectionA = RvfiDiiConnection socA socATraceVer "implementation A" (optVerbosity flags)
+  let connectionB = RvfiDiiConnection socB socBTraceVer "implementation B" (optVerbosity flags)
+
+
   addrInstr <- mapM (resolve "127.0.0.1") (instrPort flags)
   instrSoc <- mapM (open "instruction-generator-port") addrInstr
   --
   alive <- newIORef True -- Cleared when either implementation times out, since they will may not be able to respond to future queries
   let checkSingle tc verbosity doShrink len onFail = do
         quickCheckWithResult (Args Nothing 1 1 len (verbosity > 0) (if doShrink then 1000 else 0))
-                             (prop socA socB alive onFail archDesc (timeoutDelay flags) (verbosity > 1) (return tc))
+                             (prop connectionA connectionB alive onFail archDesc (timeoutDelay flags) (verbosity > 1) (return tc))
   let check_mcause_on_trap tc traceA traceB =
         if or (map rvfiIsTrap traceA) || or (map rvfiIsTrap traceB)
            then tc <> TC (const []) [TS False [encode csrrs 0x342 0 1, encode csrrs 0x343 0 1, encode csrrs 0xbc0 0 1]]
            else tc
   let saveOnFail tc tcTrans = do
         let (rawInsts, _) = fromTestCase tc
         let insts = (map diiInstruction rawInsts) ++ [diiEnd]
-        m_traces <- doRVFIDII socA socB alive (timeoutDelay flags) False insts
+        m_traces <- doRVFIDII connectionA connectionB alive (timeoutDelay flags) False insts
         let (traceA, traceB) = fromMaybe (error "unexpected doRVFIDII failure")
                                          m_traces
         writeFile "last_failure.S" ("# last failing test case:\n" ++ show tc)
-        when (optVerbosity flags > 0) $
+        when (optVerbosity flags > 0 && optShrink flags) $
           do putStrLn "Replaying shrunk failed test case:"
              let tcNew = tcTrans tc traceA traceB
-             checkSingle tcNew 2 False (testLen flags) (const $ return ())
+             _ <- checkSingle tcNew 2 False (testLen flags) (const $ return ())
```

Comment:
This should not be needed.

Reply:
Yeah not needed but fixes:
```
warning: [-Wunused-do-bind]     A do-notation statement discarded a result of type ‘Result’     Suppress this warning by saying       ‘_ <- checkSingle               tcNew 2 False (testLen flags) (const $ return ())
```

Reply:
Hm, right, because it's `IO Result` not `IO ()`. Maybe we should just be calling `prop` directly rather than going through QuickCheck again.